[REFACTOR] 마이페이지 관련 클래스 생성 및 로직 분리

liimjiin opened the pull request:
## 관련 이슈
<!-- 관련된 이슈 번호를 적어주세요. -->
<!-- ex) Closes #이슈번호 -->
- Closes #21 

## 변경 사항
<!-- 이 PR에서 변경된 주요 내용을 설명해주세요.
     변경사항이 여러 개일 경우, 주요 내용과 하위 항목을 구분하여 작성해주세요. -->
<!-- ex) 1. 000 엔티티 구현 -->
<!--        - 000 메서드 구현 -->
1. `MyPageController` 생성
   - `TaskController`의 마이페이지 관련 로직을 분리하여 `MyPageController`로 이동
   - 사용자 마이페이지 정보를 조회하는 `/api/mypage` 엔드포인트 생성
2. `MyPageService` 생성
   - `TaskService`의 마이페이지 관련 메서드를 분리하여 `MyPageService`로 이동
   - `calculateCompletionRate`메서드 접근 제한자 `public` -> `private` 변경
3. `TaskController`, `TaskService` 수정
   - 마이페이지 관련 로직 삭제
   - `MyPageResponseDto` import 삭제

## 스크린샷
<!-- (선택) 주요 변경사항(기능 구현 화면, 테스트 결과 등)에 대한 이미지를 삽입해주세요. -->
<!-- 변경사항에 따라 아래 표 템플릿을 활용해주세요 -->
`MyPage 조회 기능`
|변경 전|변경 후 |
|---|---|
|`TaskController`에서 마이페이지 조회|`MyPageController`에서 마이페이지 조회|
## 체크리스트
<!-- PR을 제출하기 전에 확인해야 할 사항들입니다. -->
- [x] 코드 컨벤션을 준수하였습니까?
- [x] 모든 테스트를 통과하였습니까?
- [x] 관련 문서를 업데이트하였습니까?

## 공유사항
<!-- 리뷰어에게 전달할 추가 정보가 있다면 여기에 적어주세요. -->
- MyPageController를 생성하여 마이페이지 엔드포인트를 분리했습니다.
- TaskController와 TaskService에서 마이페이지 관련 로직을 분리하고 테스트까지 진행 완료했습니다. 

## 추후 업무
<!-- 추후 진행될 과제에 대해 적어주세요. -->
<!-- ex) [ ] 000 기능 구현 -->
- [ ] 마이페이지 추가 기능 논의 및 개발

liimjiin committed:
REFACTOR: MyPageController 생성 및 구현
- MyPageController 클래스 생성
- 마이페이지 엔드포인트 설정
- TaskController 마이페이지 관련 로직 MyPageController로 분리

liimjiin committed:
REFACTOR: MyPageService 생성 및 구현
- MyPageService 클래스 생성
- TaskService 마이페이지 관련 메서드 MyPageService로 분리
- calculateCompletionRate 접근 제한자 public -> private 변경
- 메서드 기능 설명 주석 추가

liimjiin committed:
REFACTOR: 마이페이지 관련 로직 삭제
- MyPageResponseDto import 삭제
- 마이페이지 로직 삭제

liimjiin committed:
REFACTOR: 마이페이지 관련 메서드 삭제
- MyPageResponseDto import 삭제
- 마이페이지 관련 메서드 삭제

Chan-GN commented on src/main/java/org/example/dotoli/service/MyPageService.java:
```diff
@@ -0,0 +1,67 @@
+	/**
+	 * 달성률 계산
+	 */
+	private Long calculateCompletionRate(Long memberId) {
```
이 메서드는 불필요한 중복 쿼리를 포함하고 있습니다. 매개변수를 변경하여 중복 쿼리를 제거해보세요.

Chan-GN commented on src/main/java/org/example/dotoli/service/MyPageService.java:
```diff
@@ -0,0 +1,67 @@
+	/**
+	 * 사용자의 모든 할 일 개수 조회
+	 */
+	public Long getTotalTaskCountForMember(Long memberId) {
```
단순 조회 메서드를 메서드로 추출하는 것에 대해 좀 더 고민해보시면 좋겠습니다.

Chan-GN commented on src/main/java/org/example/dotoli/service/MyPageService.java:
```diff
@@ -0,0 +1,67 @@
+		Long completedTasks = getCompletedTaskCountForMember(memberId);
+		Long completionRate = calculateCompletionRate(memberId);
+
+		return new MyPageResponseDto(member.getEmail(), member.getNickname(), totalTasks, completedTasks,
```
여기서 Member 엔터티에 대해 불필요한 쿼리가 발생하지 않는지 확인하시면 좋을 것 같습니다.
만약 Member 엔티티에 대해 쿼리가 발생한다면 컨트롤러 레이어의 userDetails의 Member 정보를 바로 사용하는 것을 고려해보세요.

Chan-GN commented on src/main/java/org/example/dotoli/service/MyPageService.java:
```diff
@@ -0,0 +1,67 @@
+	public MyPageResponseDto getMyPageInfo(Long memberId) {
+		Member member = memberRepository.getReferenceById(memberId);
+
+		Long totalTasks = getTotalTaskCountForMember(memberId);
```
변수명을 totalTaskCount와 같이 변경하는 것을 고려해보세요.

hyerin315 replied:
찬근님 의견에 동감합니다. 
변수명을 메서드명과 같은 방향으로 Count까지 포함하는 게 직관적일 것 같기도 하네요.

liimjiin replied:
리뷰 감사드립니다! 말씀하신 부분 수정하겠습니다 :D

Chan-GN reviewed (changes requested):
전반적으로 적절하게 리팩토링을 잘해주신 것 같습니다. 고생 많으셨어요.
다만 몇 가지 수정하면 좋을 것 같은 부분이 보여 리뷰를 남깁니다.
각 라인에 남긴 리뷰를 참고하셔서 리팩토링을 다시 진행해보시면 좋겠습니다.

hyerin315 reviewed (approved):
리팩토링까지 고생 많으셨습니다!

liimjiin committed:
REFACTOR: 마이페이지 관련 로직 수정
- 불필요한 쿼리 방지를 위해 Controller에서 사용자 정보 할당
- MyPageResponseDto에 setMemberInfo 메서드 추가
- 가독성을 위한 변수명 수정

liimjiin committed:
REFACTOR: 마이페이지 관련 로직 수정
FIX: main 브랜치 문제 해결

Chan-GN reviewed (approved):
중복 및 불필요한 쿼리를 적절하게 잘 제거하신 것 같습니다.
변수명도 잘 변경해주셔서 어떤 역할을 수행하는 메서드인지 보다 쉽게 파악할 수 있었습니다.
고생 많으셨습니다!

cuteJJong reviewed (approved):
리팩토링까지 하시느라 고생많으셨습니다. 의도를 파악하기 쉽고 좋았습니다.
